bpo-43040: Fix randrange() is very slow if the range is a power of 2.

dbaarda opened the pull request:
Make _randbelow_with_getrandbits() get the number of bits needed for (n-1), not n, since we only return values <=n. When n is a power of 2, this reduces the number of bits needed by 1, and also means getrandbits() is not called repeatedly (on average 2x as many times as needed) until the returned value is <n.

Make  _randbelow_without_getrandbits() handling of n==0 consistent with _randbelow_with_getrandbits() by using `if not n:` instead of `if n==0:`.

Update test_random.py and test_generators.py to reflect changes to the random repeatability introduced by this optimization.

<!-- issue-number: [bpo-43040](https://bugs.python.org/issue43040) -->
https://bugs.python.org/issue43040
<!-- /issue-number -->

dbaarda committed:
Make _randbelow_with_getrandbits() faster when n is a power of 2.

Make _randbelow_with_getrandbits() get the number of bits needed for (n-1),
not n, since we only return values <=n. When n is a power of 2, this reduces
the number of bits needed by 1, and also means getrandbits() is not called
repeatedly (on average 2x as many times as needed) until the returned value is
<n.

Make _randbelow_with_getrandbits() and _randbelow_without_getrandbits()
short-circuit and return 0 if n <= 1, not just == 0. This also protects
against invalid evaluation in the above fix to _randbelow_with_getrandbits()
calling getrandbits(0) when n == 1.

vstinner commented on Lib/random.py:
```diff
@@ -240,10 +240,10 @@ def __init_subclass__(cls, /, **kwargs):
     def _randbelow_with_getrandbits(self, n):
         "Return a random int in the range [0,n).  Returns 0 if n==0."
```
```suggestion
        "Return a random int in the range [0,n).  Returns 0 if n<=1."
```

Same remark for _randbelow_without_getrandbits() docstring.

dbaarda replied:
Yeah, you are right. I've just pushed the updates to both docstrings.

dbaarda committed:
Fix test_random.py now randrange() doesn't call getrandbits() repeatedly.

dbaarda committed:
Update docstrings for _randbelow_with_getrandbits() and _randbelow_without_getrandbits().

These will return 0 for n <=1.

mdickinson commented on Lib/random.py:
```diff
@@ -238,19 +238,19 @@ def __init_subclass__(cls, /, **kwargs):
 
     def _randbelow_with_getrandbits(self, n):
-        "Return a random int in the range [0,n).  Returns 0 if n==0."
+        "Return a random int in the range [0,n).  Returns 0 if n<=1."
```
This change is unnecessary, and confusing. `n == 0` is a genuine special case that it makes sense to call out in the docstring. `n == 1` is not a special case, and the behaviour for `n == 1` is already covered by the "Return a random int in the range [0, n)" part of the docstring.

dbaarda replied:
Ah.. OK, I understand.

mdickinson commented on Lib/random.py:
```diff
@@ -238,19 +238,19 @@ def __init_subclass__(cls, /, **kwargs):
+        "Return a random int in the range [0,n).  Returns 0 if n<=1."
 
-        if not n:
+        if n <= 1:
```
This change is also unnecessary, and likely slows down the common case a tiny amount. `getrandbits(0)` works in master.

dbaarda replied:
I just tested this for python3.9 on Debian testing and `if not n:` is indeed faster than the alternatives here;

```
$ python3 -m timeit -s 'i = 1234' 'if not i: pass'
10000000 loops, best of 5: 33.7 nsec per loop

$ python3 -m timeit -s 'i = 1234' 'if i==0: pass'
5000000 loops, best of 5: 56.5 nsec per loop

$ python3 -m timeit -s 'i = 1234' 'if i<=1: pass'
5000000 loops, best of 5: 59.2 nsec per loop
```
I'll change this back and run tests to ensure getrandbits(0) does indeed handle this now.

mdickinson commented on Lib/random.py:
```diff
@@ -238,19 +238,19 @@ def __init_subclass__(cls, /, **kwargs):
 
     def _randbelow_without_getrandbits(self, n, maxsize=1<<BPF):
-        """Return a random int in the range [0,n).  Returns 0 if n==0.
+        """Return a random int in the range [0,n).  Returns 0 if n<=1.
```
Suggest reverting this change; there's no need to change this docstring.

mdickinson commented on Lib/random.py:
```diff
@@ -261,7 +261,7 @@ def _randbelow_without_getrandbits(self, n, maxsize=1<<BPF):
                 "To remove the range limitation, add a getrandbits() method.")
             return _floor(random() * n)
-        if n == 0:
+        if n <= 1:
```
Similarly, I don't see the value of this change.

dbaarda committed:
Revert "Update docstrings for _randbelow_with_getrandbits() and _randbelow_without_getrandbits()."

This reverts commit 5215ba385543646897b577f77c055e02f46ed2c5.

The docstring already explains the behaviour when n==1 with the normal
behaviour of returning [0,n). The case when n=0 is special and needs specific
explanation.

dbaarda committed:
Use `if not n:` instead of `if n<=1:` in _randbelow_*_getrandbits().

The case of n==1 is correctly handled by the rest of the code, and this makes
it a tiny bit faster for the common case (n > 1).

This also makes how this case is handled in _randbelow_without_getrandbits()
consistent with _randbelow_with_getrandbits().

dbaarda committed:
Fix test_generators.py for repeatability change introduced.

This test uses gen.choice() to randomly select values from a sequence, which
is affected by the repeatability change from optimizing _randbelow_with_getrandbits().

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
